server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -302,6 +312,10 @@ public SearchService(
 
         lowLevelCancellation = LOW_LEVEL_CANCELLATION_SETTING.get(settings);
         clusterService.getClusterSettings().addSettingsUpdateConsumer(LOW_LEVEL_CANCELLATION_SETTING, this::setLowLevelCancellation);
+
+        BooleanQuery.setMaxClauseCount(INDICES_MAX_CLAUSE_COUNT_SETTING.get(settings));
```

Comment:
I have the concerns regarding the impact of this change. First of all, to acknowledge, making this property adjustable is useful. However implementation-wise, since `BooleanQuery::setMaxClauseCount` is static setting, changing it could impact the queries which are already being submitted / executed in a very weird ways. Any thoughts on that?

Reply:
the limits get checked at the time of query parse/rewrite - so at shard level, the query will either succeed or fail on the limit of clauses

Here are different scenarios:

Initial State: with limit at 1024
* Query with 1500 clauses will be rejected

Transition state: limit raised to 2048 via the dynamic setting added here
* Ongoing/queries submitted concurrent with the settings change (say to raise the threshold) will initially fail with too many clauses (as before) 
* Eventual consistency will prevail and all nodes will eventually apply the updated setting(within state lag period)
* In the case of a query spanning multiple shards across nodes,  there could be scenario where shards on the node that have applied settings respond successfully and shards on nodes with old setting will return failure and such queries will see partial success. 
* Eventually post applying of the setting, all queries with clauses < 2048 will get passed

Does this address your concerns? If not, it will help to share the exact scenario you had in mind

Reply:
Thanks @malpani , this "... eventual consistency ..." part is  essentially what bothers me, not only between nodes but between multiple shards on the same node: it may happen that for the same search request, while it is being distributed between shards, some shards may see old setting whereas others may see new one.

Not sure it is solvable at all, unless we take into account the state of search-related thread pools and delay the change, probably adding documentation notes about the impact of changing this setting on clusters would be sufficient.   

Thank you.

Reply:
ack, I will tag you on the doc update PR. 

Intra-node consistency probably can be achieved by making it volatile in Lucene. We would still have the inter node eventual consistency so will document this clearly.

Reply:
Is there another example where we are adjusting a dynamic property in this way or is it the first time? I'm all for YOLOing a setting at runtime, but maybe @nknize has a stronger opinion on this?

Reply:
Haven't dug deeper here but thinking out loud, if we could do such check on the coordinator node that spans out the search request, we could probably avoid intra and inter node inconsistencies by rejecting at the coordinator itself.

Reply:
@dblock - yes, this is a common approach for updating dynamic settings for example - https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/SearchService.java#L294 . the slight uniqueness here is it is updating a static setting.

@setiah - Interesting thought - while this clause validation can be done on coordinator side, I dont want to parse yet again for checking on this failure scenario and affect the happy paths. Background - each shard query does a rewrite based on local stats for that shard (not available on coordinator) - so the per shard rewrite is not avoidable today. One of the long term exploration areas is to do better coordinator side planning and query estimation and this could be revisited then. I want to keep this change simple and dont really see a problem with eventual consistency semantics here (this is inline with other setting updates)